File: CucumberJVM/Chapter13/13.1-TheLocalScenarioContextPattern/Before/src/main/java/com/wimp/localscenariocontext/app/services/OrderService.java
package com.wimp.localscenariocontext.app.services;

import com.wimp.localscenariocontext.app.infrastructure.DataContext;
import com.wimp.localscenariocontext.app.models.Order;
import com.wimp.localscenariocontext.app.models.OrderStatus;

public class OrderService {
    public static void placeOrder(String customerName, int orderNo, String pizzaName) {
        if (!AuthenticationService.isLoggedIn(customerName)) {
            throw new RuntimeException("Customer " + customerName + " is not logged in.");
        }

        Order order = new Order(orderNo, customerName, pizzaName);
        setStatus(order, OrderStatus.PLACED);
        DataContext.INSTANCE.saveOrder(order);
    }

    private static void setStatus(Order order, OrderStatus status) {
        order.setStatus(status);
    }

    public static void cancelOrder(String customerName, int orderNo) {
        Order order = DataContext.INSTANCE.getOrderByOrderNr(orderNo);
        if (order == null) {
            throw new RuntimeException("Order " + orderNo + " does not exist.");
        }

        if (!order.getCustomerName().equals(customerName)) {
            throw new RuntimeException("Order " + orderNo + " belongs to " + order.getCustomerName() + ", not " + customerName + ".");
        }

        NotificationService.sendCancellationNotification(customerName);
        order.setStatus(OrderStatus.CANCELLED);
    }
}

File: Reqnroll/Chapter13/13.1-TheLocalScenarioContextPattern/After/WIMP.LocalScenarioContextSample.App/Services/OrderService.cs
using WIMP.LocalScenarioContextSample.App.Infrastructure;
using WIMP.LocalScenarioContextSample.App.Models;

namespace WIMP.LocalScenarioContextSample.App.Services;

public static class OrderService
{
    public static void PlaceOrder(string? customerName, int orderNo, string pizzaName)
    {
        if (customerName == null || !AuthenticationService.IsLoggedIn(customerName))
        {
            throw new InvalidOperationException($"Customer '{customerName}' is not logged in.");
        }

        var order = new Order(orderNo, customerName, pizzaName);
        SetStatus(order, OrderStatus.Placed);
        DataContext.Instance.SaveOrder(order);
    }

    private static void SetStatus(Order order, OrderStatus status)
    {
        order.Status = status;
    }

    public static void CancelOrder(string? customerName, int orderNo)
    {
        var order = DataContext.Instance.GetOrderByOrderNr(orderNo) ??
            throw new InvalidOperationException($"Order {orderNo} does not exist.");

        if (order.CustomerName != customerName)
        {
            throw new InvalidOperationException($"Order {orderNo} belongs to {order.CustomerName}, not {customerName}.");
        }

        NotificationService.SendCancellationNotification(customerName);
        order.Status = OrderStatus.Cancelled;
    }
}
